include/ghex/transport_layer/util/pmi/pmix/pmi.hpp

```diff
@@ -23,133 +23,160 @@ namespace gridtools
-			ERR("Client ns %s rank %d: PMIx_Get %s: got wrong data type\n", myproc.nspace, myproc.rank, key.c_str());
-		    }
-
-		    /* get the returned data */
-		    std::vector<char> data(pvalue->data.bo.bytes, pvalue->data.bo.bytes+pvalue->data.bo.size);
-
-		    /* free the PMIx data */
-		    PMIX_VALUE_RELEASE(pvalue);
-		    PMIX_PROC_DESTRUCT(&proc);
-
-		    LOG("PMIx_get %s returned %zi bytes", key.c_str(), data.size());
-		    
-		    return  data;
-		}
-	    };
-	} // namespace tl
+        namespace tl
+        {
+            typedef struct {
+                pthread_mutex_t mutex;
+                pthread_cond_t  cond;
+                pmix_status_t   status;
+                volatile bool   active;
+            } cblock_t;
+
+            template<>
+            class pmi<pmix_tag>
+            {
+
+            private:
+                pmix_proc_t allproc;
+                pmix_proc_t myproc;
+                int32_t nprocs;
+
+            public:
+                using rank_type = int;
+                using size_type = int;
+
+            public:
+
+                pmi()
+                {
+                    int rc;
+                    pmix_value_t *pvalue;
+
+                    if (PMIX_SUCCESS != (rc = PMIx_Init(&myproc, NULL, 0))) {
+                        ERR("PMIx_Init failed");
+                        exit(1);
+                    }
+                    if(myproc.rank == 0) LOG("PMIx initialized");
+
+                    /* job-related info is found in our nspace, assigned to the
+                     * wildcard rank as it doesn't relate to a specific rank. Setup
+                     * a name to retrieve such values */
+                    PMIX_PROC_CONSTRUCT(&allproc);
+                    // (void)strncpy(allproc.nspace, myproc.nspace, PMIX_MAX_NSLEN);
+                    (void)memcpy(allproc.nspace, myproc.nspace, PMIX_MAX_NSLEN);
+                    allproc.rank = PMIX_RANK_WILDCARD;
+
+                    /* get the number of procs in our job */
+                    if (PMIX_SUCCESS != (rc = PMIx_Get(&allproc, PMIX_JOB_SIZE, NULL, 0, &pvalue))) {
+                        std::string nspace(myproc.nspace, myproc.nspace+strlen(myproc.nspace));
+                        throw std::string("Client ns " + nspace + " rank " + std::to_string(myproc.rank) + 
+                            ": PMIx_Get job size failed: " + std::to_string(rc) + "\n");
+                        exit(1);
+                    }
+                    nprocs = pvalue->data.uint32;
+                    PMIX_VALUE_RELEASE(pvalue);
+                }
+
+                ~pmi()
+                {
+                    int rc;
+                    if (PMIX_SUCCESS != (rc = PMIx_Finalize(NULL, 0))) {
+                        WARN("Client ns %s rank %d:PMIx_Finalize failed: %d\n", myproc.nspace, myproc.rank, rc);
+                    } else {
+                        if(myproc.rank == 0) LOG("PMIx finalized");
+                    }
+                }
```

Comment:
this class manages a resource (`pmix`) and thus it should either be **_non-copyable_** (move-only) or hold the resource in a `shared_ptr`

Reply:
Please have a look if this is what you had in mind. I only check for moved bit in the destructor though. Seems to me like it is not needed elsewhere, but maybe I'm wrong.